h2o-core/src/main/java/water/Scope.java

```diff
@@ -21,34 +21,49 @@
 
 public class Scope {
   // Thread-based Key lifetime tracking
-  static private final ThreadLocal<Scope> _scope = new ThreadLocal<Scope>() {
+  private static final ThreadLocal<Scope> _scope = new ThreadLocal<Scope>() {
     @Override protected Scope initialValue() { return new Scope(); }
   };
-  private final Stack<HashSet<Key>> _keys = new Stack<>();
+  private final Stack<Set<Key>> _keys = new Stack<>();
+  
+  private final Stack<Set<Key>> _protectedKeys = new Stack<>();
   
   /** debugging purpose */
-  static public Scope current() {
+  public static Scope current() {
     return _scope.get();
   }
 
   /** Enter a new Scope */
-  static public void enter() { _scope.get()._keys.push(new HashSet<Key>()); }
+  public static void enter() { 
+    Scope scope = _scope.get();
+    scope._keys.push(new HashSet<>()); 
+    Set<Key> outerProtected = scope._protectedKeys.empty() ? Collections.emptySet() : scope._protectedKeys.peek(); 
+    scope._protectedKeys.push(new HashSet<>(outerProtected)); //inherit protected keys from outer scope
+  }
 
-  /** Exit the inner-most Scope, remove all Keys created since the matching
+  /** Exit the innermost Scope, remove all Keys created since the matching
    *  enter call except for the listed Keys.
    *  @return Returns the list of kept keys. */
-  static public Key[] exit(Key... keep) {
-    List<Key> keylist = new ArrayList<>();
-    if( keep != null )
-      for( Key k : keep ) if (k != null) keylist.add(k);
-    Object[] arrkeep = keylist.toArray();
+  public static Key[] exit(Key... keep) {
+    Set<Key> keepKeys = new HashSet<>();
+    if (keep != null) {
+      for (Key k : keep) {
+        if (k != null) keepKeys.add(k);
+      }
+    }
+    Scope scope = _scope.get();
+    keepKeys.addAll(scope._protectedKeys.pop());
+    Key[] arrkeep = keepKeys.toArray(new Key[0]);
     Arrays.sort(arrkeep);
-    Stack<HashSet<Key>> keys = _scope.get()._keys;
-    if (keys.size() > 0) {
+    Stack<Set<Key>> removeKeys = scope._keys;
+    if (!removeKeys.empty()) {
       Futures fs = new Futures();
-      for (Key key : keys.pop()) {
-        int found = Arrays.binarySearch(arrkeep, key);
-        if (arrkeep.length == 0 || found < 0) Keyed.remove(key, fs, true);
+      for (Key key : removeKeys.pop()) {
+        boolean remove = arrkeep.length == 0 || Arrays.binarySearch(arrkeep, key) < 0;
+        if (remove) {
+          boolean cascade = !(key.get() instanceof Frame); //Frames are handled differently as we're explicitly also tracking their Vec keys...
```

Comment:
not very elegant, but the fact that Vecs were deleted in 2 different ways is annoying: (1) as tracked keys from a tracked frame, (2) as a direct component of the frame itself.
Also see the changes in Frame class, where vecs are now not fully deleted if `cascade=false`.